File: backend/api/recovery_readiness.go
package api

import (
	"net/http"
	"strings"

	"go-stock/backend/db"
	"go-stock/backend/recovery"

	"gorm.io/gorm"
)

// RecoveryReadinessView is the snake_case HTTP DTO for
// GET /api/recovery/readiness. It intentionally exposes no persistence model.
type RecoveryReadinessView struct {
	Status             string                         `json:"status"`
	CheckpointStatus   RecoveryCheckpointStatusView   `json:"checkpoint_status"`
	AuditContinuity    RecoveryAuditContinuityView    `json:"audit_continuity"`
	ReplayVerification RecoveryReplayVerificationView `json:"replay_verification"`
	DivergenceSummary  RecoveryDivergenceSummaryView  `json:"divergence_summary"`
}

type RecoveryCheckpointStatusView struct {
	ScopeCount            int                            `json:"scope_count"`
	IntactCount           int                            `json:"intact_count"`
	CorruptCount          int                            `json:"corrupt_count"`
	ContractMismatchCount int                            `json:"contract_mismatch_count"`
	MissingBaseline       bool                           `json:"missing_baseline"`
	MaxLagEvents          uint                           `json:"max_lag_events"`
	Samples               []RecoveryCheckpointSampleView `json:"samples"`
	StatusHint            string                         `json:"status_hint"`
}

type RecoveryCheckpointSampleView struct {
	Scope       string `json:"scope"`
	LastAuditID uint   `json:"last_audit_id"`
	LastEventID string `json:"last_event_id"`
	Version     int    `json:"version"`
	Intact      bool   `json:"intact"`
	LagEvents   uint   `json:"lag_events"`
	Error       string `json:"error,omitempty"`
}

type RecoveryAuditContinuityView struct {
	MaxAuditID     uint   `json:"max_audit_id"`
	EventCount     int64  `json:"event_count"`
	HighWatermark  uint   `json:"high_watermark"`
	WindowChecked  bool   `json:"window_checked"`
	GapCount       int    `json:"gap_count"`
	AnchorMismatch int    `json:"anchor_mismatch"`
	StatusHint     string `json:"status_hint"`
}

type RecoveryReplayVerificationView struct {
	Mode         string `json:"mode"`
	SampleSize   int    `json:"sample_size"`
	PassedCount  int    `json:"passed_count"`
	FailedCount  int    `json:"failed_count"`
	SkippedCount int    `json:"skipped_count"`
	LastError    string `json:"last_error,omitempty"`
	StatusHint   string `json:"status_hint"`
}

type RecoveryDivergenceSummaryView struct {
	Total         int                          `json:"total"`
	ByKind        map[string]int               `json:"by_kind"`
	BySeverity    map[string]int               `json:"by_severity"`
	BlockingCount int                          `json:"blocking_count"`
	Top           []RecoveryDivergenceItemView `json:"top"`
}

type RecoveryDivergenceItemView struct {
	Kind      string `json:"kind"`
	Severity  string `json:"severity"`
	Blocking  bool   `json:"blocking"`
	Scope     string `json:"scope,omitempty"`
	EventID   string `json:"event_id,omitempty"`
	EventType string `json:"event_type,omitempty"`
	Detail    string `json:"detail,omitempty"`
}

type recoveryReadinessEvaluator func(*gorm.DB, *recovery.Options) recovery.RecoveryReadinessResult

// RecoveryReadinessHandler is a read-only visibility adapter. It delegates all
// rules to recovery.EvaluateRecoveryReadiness and only maps the returned DTO.
type RecoveryReadinessHandler struct {
	database *gorm.DB
	evaluate recoveryReadinessEvaluator
}

// NewRecoveryReadinessHandler creates the production handler.
func NewRecoveryReadinessHandler(database *gorm.DB) *RecoveryReadinessHandler {
	if database == nil {
		database = db.Dao
	}
	return &RecoveryReadinessHandler{
		database: database,
		evaluate: recovery.EvaluateRecoveryReadiness,
	}
}

// NewRecoveryReadinessHandlerWithEvaluator creates an injectable handler for
// isolated API tests. Production callers should use NewRecoveryReadinessHandler.
func NewRecoveryReadinessHandlerWithEvaluator(
	database *gorm.DB,
	evaluate func(*gorm.DB, *recovery.Options) recovery.RecoveryReadinessResult,
) *RecoveryReadinessHandler {
	h := NewRecoveryReadinessHandler(database)
	if evaluate != nil {
		h.evaluate = evaluate
	}
	return h
}

// ServeHTTP serves GET /api/recovery/readiness only.
func (h *RecoveryReadinessHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if h == nil {
		h = NewRecoveryReadinessHandler(nil)
	}
	path := strings.TrimSuffix(r.URL.Path, "/")
	if path != "/api/recovery/readiness" {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
		return
	}
	if r.Method != http.MethodGet {
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
		return
	}
	if h.evaluate == nil {
		h.evaluate = recovery.EvaluateRecoveryReadiness
	}
	result := h.evaluate(h.database, nil)
	writeJSON(w, http.StatusOK, mapRecoveryReadinessView(result))
}

func mapRecoveryReadinessView(result recovery.RecoveryReadinessResult) RecoveryReadinessView {
	samples := make([]RecoveryCheckpointSampleView, 0, len(result.CheckpointStatus.Samples))
	for _, sample := range result.CheckpointStatus.Samples {
		samples = append(samples, RecoveryCheckpointSampleView{
			Scope:       sample.Scope,
			LastAuditID: sample.LastAuditID,
			LastEventID: sample.LastEventID,
			Version:     sample.Version,
			Intact:      sample.Intact,
			LagEvents:   sample.LagEvents,
			Error:       sample.Error,
		})
	}

	top := make([]RecoveryDivergenceItemView, 0, len(result.DivergenceSummary.Top))
	for _, item := range result.DivergenceSummary.Top {
		top = append(top, RecoveryDivergenceItemView{
			Kind:      item.Kind,
			Severity:  item.Severity,
			Blocking:  item.Blocking,
			Scope:     item.Scope,
			EventID:   item.EventID,
			EventType: item.EventType,
			Detail:    item.Detail,
		})
	}

	byKind := result.DivergenceSummary.ByKind
	if byKind == nil {
		byKind = map[string]int{}
	}
	bySeverity := result.DivergenceSummary.BySeverity
	if bySeverity == nil {
		bySeverity = map[string]int{}
	}

	return RecoveryReadinessView{
		Status: result.Status,
		CheckpointStatus: RecoveryCheckpointStatusView{
			ScopeCount:            result.CheckpointStatus.ScopeCount,
			IntactCount:           result.CheckpointStatus.IntactCount,
			CorruptCount:          result.CheckpointStatus.CorruptCount,
			ContractMismatchCount: result.CheckpointStatus.ContractMismatchCount,
			MissingBaseline:       result.CheckpointStatus.MissingBaseline,
			MaxLagEvents:          result.CheckpointStatus.MaxLagEvents,
			Samples:               samples,
			StatusHint:            result.CheckpointStatus.StatusHint,
		},
		AuditContinuity: RecoveryAuditContinuityView{
			MaxAuditID:     result.AuditContinuity.MaxAuditID,
			EventCount:     result.AuditContinuity.EventCount,
			HighWatermark:  result.AuditContinuity.HighWatermark,
			WindowChecked:  result.AuditContinuity.WindowChecked,
			GapCount:       result.AuditContinuity.GapCount,
			AnchorMismatch: result.AuditContinuity.AnchorMismatch,
			StatusHint:     result.AuditContinuity.StatusHint,
		},
		ReplayVerification: RecoveryReplayVerificationView{
			Mode:         result.ReplayVerification.Mode,
			SampleSize:   result.ReplayVerification.SampleSize,
			PassedCount:  result.ReplayVerification.PassedCount,
			FailedCount:  result.ReplayVerification.FailedCount,
			SkippedCount: result.ReplayVerification.SkippedCount,
			LastError:    result.ReplayVerification.LastError,
			StatusHint:   result.ReplayVerification.StatusHint,
		},
		DivergenceSummary: RecoveryDivergenceSummaryView{
			Total:         result.DivergenceSummary.Total,
			ByKind:        byKind,
			BySeverity:    bySeverity,
			BlockingCount: result.DivergenceSummary.BlockingCount,
			Top:           top,
		},
	}
}

// RegisterRecoveryReadinessRoutes mounts the read-only route (httptest/server).
func RegisterRecoveryReadinessRoutes(mux *http.ServeMux) {
	mux.Handle("/api/recovery/readiness", NewRecoveryReadinessHandler(nil))
}

// RegisterRecoveryReadinessHandler mounts an injected handler for tests.
func RegisterRecoveryReadinessHandler(mux *http.ServeMux, h *RecoveryReadinessHandler) {
	if h == nil {
		h = NewRecoveryReadinessHandler(nil)
	}
	mux.Handle("/api/recovery/readiness", h)
}

// RecoveryReadinessAssetMiddleware mounts the visibility route on Wails.
func RecoveryReadinessAssetMiddleware(next http.Handler) http.Handler {
	h := NewRecoveryReadinessHandler(nil)
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if strings.HasPrefix(r.URL.Path, "/api/recovery/") {
			h.ServeHTTP(w, r)
			return
		}
		next.ServeHTTP(w, r)
	})
}

File: frontend/src/api/recoveryReadiness.ts
/** Recovery Readiness 只读 HTTP API（Phase6.5.8.5.3） */

export type RecoveryCheckpointSample = {
  scope: string
  last_audit_id: number
  last_event_id: string
  version: number
  intact: boolean
  lag_events: number
  error?: string
}

export type RecoveryCheckpointStatus = {
  scope_count: number
  intact_count: number
  corrupt_count: number
  contract_mismatch_count: number
  missing_baseline: boolean
  max_lag_events: number
  samples: RecoveryCheckpointSample[]
  status_hint: string
}

export type RecoveryAuditContinuity = {
  max_audit_id: number
  event_count: number
  high_watermark: number
  window_checked: boolean
  gap_count: number
  anchor_mismatch: number
  status_hint: string
}

export type RecoveryReplayVerification = {
  mode: string
  sample_size: number
  passed_count: number
  failed_count: number
  skipped_count: number
  last_error?: string
  status_hint: string
}

export type RecoveryDivergenceItem = {
  kind: string
  severity: string
  blocking: boolean
  scope?: string
  event_id?: string
  event_type?: string
  detail?: string
}

export type RecoveryDivergenceSummary = {
  total: number
  by_kind: Record<string, number>
  by_severity: Record<string, number>
  blocking_count: number
  top: RecoveryDivergenceItem[]
}

export type RecoveryReadinessView = {
  status: string
  checkpoint_status: RecoveryCheckpointStatus
  audit_continuity: RecoveryAuditContinuity
  replay_verification: RecoveryReplayVerification
  divergence_summary: RecoveryDivergenceSummary
}

function asRecord(v: unknown): Record<string, unknown> {
  return v && typeof v === 'object' ? (v as Record<string, unknown>) : {}
}

function mapSample(raw: unknown): RecoveryCheckpointSample {
  const s = asRecord(raw)
  return {
    scope: String(s.scope || ''),
    last_audit_id: Number(s.last_audit_id) || 0,
    last_event_id: String(s.last_event_id || ''),
    version: Number(s.version) || 0,
    intact: !!s.intact,
    lag_events: Number(s.lag_events) || 0,
    error: s.error ? String(s.error) : '',
  }
}

function mapDivergenceItem(raw: unknown): RecoveryDivergenceItem {
  const d = asRecord(raw)
  return {
    kind: String(d.kind || ''),
    severity: String(d.severity || ''),
    blocking: !!d.blocking,
    scope: d.scope ? String(d.scope) : '',
    event_id: d.event_id ? String(d.event_id) : '',
    event_type: d.event_type ? String(d.event_type) : '',
    detail: d.detail ? String(d.detail) : '',
  }
}

function mapNumberMap(raw: unknown): Record<string, number> {
  const src = asRecord(raw)
  const out: Record<string, number> = {}
  for (const [k, v] of Object.entries(src)) {
    out[k] = Number(v) || 0
  }
  return out
}

function mapView(raw: Record<string, unknown> | null | undefined): RecoveryReadinessView | null {
  if (!raw || typeof raw !== 'object') return null
  const cp = asRecord(raw.checkpoint_status)
  const cont = asRecord(raw.audit_continuity)
  const replay = asRecord(raw.replay_verification)
  const div = asRecord(raw.divergence_summary)
  const samplesRaw = Array.isArray(cp.samples) ? cp.samples : []
  const topRaw = Array.isArray(div.top) ? div.top : []
  return {
    status: String(raw.status || ''),
    checkpoint_status: {
      scope_count: Number(cp.scope_count) || 0,
      intact_count: Number(cp.intact_count) || 0,
      corrupt_count: Number(cp.corrupt_count) || 0,
      contract_mismatch_count: Number(cp.contract_mismatch_count) || 0,
      missing_baseline: !!cp.missing_baseline,
      max_lag_events: Number(cp.max_lag_events) || 0,
      samples: samplesRaw.map(mapSample),
      status_hint: String(cp.status_hint || ''),
    },
    audit_continuity: {
      max_audit_id: Number(cont.max_audit_id) || 0,
      event_count: Number(cont.event_count) || 0,
      high_watermark: Number(cont.high_watermark) || 0,
      window_checked: !!cont.window_checked,
      gap_count: Number(cont.gap_count) || 0,
      anchor_mismatch: Number(cont.anchor_mismatch) || 0,
      status_hint: String(cont.status_hint || ''),
    },
    replay_verification: {
      mode: String(replay.mode || ''),
      sample_size: Number(replay.sample_size) || 0,
      passed_count: Number(replay.passed_count) || 0,
      failed_count: Number(replay.failed_count) || 0,
      skipped_count: Number(replay.skipped_count) || 0,
      last_error: replay.last_error ? String(replay.last_error) : '',
      status_hint: String(replay.status_hint || ''),
    },
    divergence_summary: {
      total: Number(div.total) || 0,
      by_kind: mapNumberMap(div.by_kind),
      by_severity: mapNumberMap(div.by_severity),
      blocking_count: Number(div.blocking_count) || 0,
      top: topRaw.map(mapDivergenceItem),
    },
  }
}

/**
 * GET /api/recovery/readiness
 * 直接返回 RecoveryReadinessView（无 code/ok 信封）。
 */
export async function getRecoveryReadiness(): Promise<RecoveryReadinessView> {
  const res = await fetch('/api/recovery/readiness')
  if (!res.ok) {
    throw new Error(`Recovery Readiness 请求失败: HTTP ${res.status}`)
  }
  const body = await res.json()
  const view = mapView(body)
  if (!view || !view.status) {
    throw new Error('Recovery Readiness 响应无效')
  }
  return view
}
